src/densenet.jl

```diff
@@ -1,92 +1,188 @@
-        ls[i][1].β .= weights["conv$(i ÷ 2)_blk/bn_b_0"]
-        ls[i][1].γ .= weights["conv$(i ÷ 2)_blk/bn_w_0"]
-    end
-    ls[end - 1].W .= transpose(dropdims(weights["fc6_w_0"], dims = (1, 2))) # Dense Layers
-    ls[end - 1].b .= weights["fc6_b_0"]
-    return ls
+"""
+    densenet(nblocks; growth_rate=32, reduction=0.5, num_classes=1000)
+
+Create a DenseNet model
+([reference](https://arxiv.org/abs/1608.06993)).
+
+# Arguments
+- `nblocks`: number of dense blocks between transitions
+- `growth_rate`: the output feature map growth rate of dense blocks (i.e. `k` in the paper)
+- `reduction`: the factor by which the number of feature maps is scaled across each transition
+- `nclasses`: the number of output classes
+"""
+function densenet(nblocks; growth_rate=32, reduction=0.5, nclasses=1000)
+  num_planes = 2 * growth_rate
+  layers = []
+  append!(layers, conv_bn((7, 7), 3, num_planes; stride=2, pad=(3, 3), usebias=false))
+  push!(layers, MaxPool((3, 3), stride=2, pad=(1, 1)))
+
+  for i in 1:3
+    append!(layers, dense_block(num_planes, growth_rate, nblocks[i]))
+    num_planes += nblocks[i] * growth_rate
+    out_planes = Int(floor(num_planes * reduction))
+    append!(layers, transition(num_planes, out_planes))
+    num_planes = out_planes
+  end
+
+  append!(layers, dense_block(num_planes, growth_rate, nblocks[4]))
+  num_planes += nblocks[4] * growth_rate
+  push!(layers, BatchNorm(num_planes, relu))
+
+  return Chain(layers...,
+               AdaptiveMeanPool((1, 1)),
+               flatten,
+               Dense(num_planes, nclasses))
 end
 
-struct DenseNet <: ClassificationModel{ImageNet.ImageNet1k}
-    layers::Chain
+"""
+    DenseNet(nblocks::NTuple{N, <:Integer};
+             growth_rate=32, reduction=0.5, nclasses=1000)
+
+Create a DenseNet model
+([reference](https://arxiv.org/abs/1608.06993)).
+See also [`densenet`](#).
+
+# Arguments
+- `nblocks`: number of dense blocks between transitions
+- `growth_rate`: the output feature map growth rate of dense blocks (i.e. `k` in the paper)
+- `reduction`: the factor by which the number of feature maps is scaled across each transition
+- `nclasses`: the number of output classes
+"""
+struct DenseNet{T}
+  layers::T
 end
 
-DenseNet() = DenseNet(densenet_layers())
+function DenseNet(nblocks::NTuple{N, <:Integer};
+                  growth_rate=32, reduction=0.5, nclasses=1000) where N
+  layers = densenet(nblocks; growth_rate=growth_rate,
+                             reduction=reduction,
+                             nclasses=nclasses)
 
-Base.show(io::IO, ::DenseNet) = print(io, "DenseNet()")
+  DenseNet(layers)
+end
 
 @functor DenseNet
 
 (m::DenseNet)(x) = m.layers(x)
+
+"""
+    DenseNet121(; pretrain=false)
+
+Create a DenseNet-121 model
+([reference](https://arxiv.org/abs/1608.06993)).
+Set `pretrain=true` to load the model with pre-trained weights for ImageNet.
+
+See also [`Metalhead.DenseNet`](#).
+"""
+function DenseNet121(; pretrain=false)
```

Comment:
The nicer API would be to take an argument to specify the size of model one wants. It would be more dynamic as well about what models we produce and should help consolidate some of the code.

Reply:
That would be `DenseNet` (with no number) above. I don't think `DenseNet(121)` makes much sense cause it's not clear what `DenseNet(83)` should be. If the number argument is constrained to 121, 161, 169, and 201, then it seems cleaner to put it in the name that to take it as an argument and throw an error when it doesn't match the 5 pre-defined integers.

Once we sort out the `nblocks`/`growth_rate` issue, something like `DenseNet(...)` above seems like the clearest API for the user to specify the structure they want.